### ISSUE TITLE
Anonymous inner class and @MonotonicNonNull

### ISSUE THREAD

**Issue by GoogleCodeExporter** [typetools/checker-framework#294]
```
Not sure if this is a bug or a misunderstanding on my part, but I think I'm 
getting a false positive.  Using Checker Framework v1.7.1 on Java 1.7.0_25, on 
Mac OS X 10.8.5.

    /*>>> import checkers.nullness.quals.MonotonicNonNull; */

    abstract class Monotonic
    {
        public /*@MonotonicNonNull*/String header;

        public void f()
        {
            if (header == null) {
                write("null");
            }
            else {
                runWithRetry(new Runnable() {
                    public void run() {
                        write(header);  // <-- error here
                    }
                });
            }
        }

        public abstract void write(String s);
        public abstract void runWithRetry(Runnable r);
    }

I expected no errors, but I got one:

# java -jar binary/checkers.jar -processor checkers.nullness.NullnessChecker 
Monotonic.java
Monotonic.java:15: error: incompatible types in argument.
                    write(header);
                          ^
  found   : @Initialized @MonotonicNonNull String
  required: @Initialized @NonNull String
1 error

If I replace the runWithRetry and the anonymous inner class with a simple 
"write(header)", there are no errors.

```

Original issue reported on code.google.com by `kan...@cakoose.com` on 18 Dec 2013 at 1:39

**Comment by mernst** [comment]
This is still a problem in Checker Framework version 2.3.2.
